# DATA 1 Practical 2 - Questions

Simos Gerasimou


## Classic Cars & Co

Classic Cars & Co is a UK company that has a large collection of classic cars from the 1980s. 

DataVision (the company you are working as a Data Scientist) has been contracted to analyse the data available for the cars and provide insights by analysing the different characteristics of the cars (e.g., speed, price). 

This Jupyter Notebook will be presented to the Classic Cars & Co main stakeholders who have limited knowledge about data science. So, your findings should be complemented by a suitable justification explaining what you observe and, when applicable, what does this observation mean and, possibly, why it occurs. The analysis along with the explanation will help them to understand whether they need to invest more to expand their collection.

### **Important Information**

(1) To answer these exercises, you **must first read Chapter 2: Introduction to NumPy from the Python Data Science Handbook** (https://jakevdp.github.io/PythonDataScienceHandbook/02.00-introduction-to-numpy.html)


(2) For each question (task) a description is provided accompanied (most of the time) by two cells: one for writing the Python code and another for providing the justification. Feel free to add more cells if you feel they are needed, but keep the cells corresponding to the same question close by.

**Hint 1**: If you find difficulties in solving a task, look at Chapter 2 from the Python Data Science Handbook.

**Hint 2**: Solving each task using NumPy should require less than 10 lines of code

#### **T1) Explore the dataset and for each column write its name, data type (categorical/numerical - nominal,ordinal,discrete,continuous) and its meaning (i.e., what does it capture?)**

* You may want to open the CSV file using a text editor (e.g., Notepad) or a spreadsheet editor (e.g., Excel)

**Write your answer here (the first is given)**
* Make: Categorical (Nominal) - The model of the car
* ....

### 1) Reading dataset

The classic cars dataset is available on VLE (look for classicCars.csv in the Practicals section)

In [1]:
#Using NumPy to read the dataset
import numpy as np
#Define the path to the dataset
data_path = "ClassicCars.csv"
#Define the type of each dataset column. 
#This is needed because NumPy arrays cannot directly read files with different data types
#Hence, we are using Structured arrays. 
#But, we will soon move to Pandas which makes data manipulation easier
types = ['U20', 'U10', 'U5', 'U20', 'U3', 'f4', 'f4', 'f4', 'f4', 'U10', 'i4', 'i4', 'i4', 'i4', 'i4']
#Read the dataset
data = np.genfromtxt(data_path, dtype=types, delimiter=',', names=True)

**Structured Arrays**
* Read more about structured arrays:
  * https://jakevdp.github.io/PythonDataScienceHandbook/02.09-structured-data-numpy.html
  * https://numpy.org/doc/stable/user/basics.rec.html

### Analysing the dataset


#### **Extracting the column names**

In [2]:
data.dtype.names

('make',
 'fueltype',
 'numofdoors',
 'bodystyle',
 'drivewheels',
 'wheelbase',
 'length',
 'width',
 'height',
 'numofcylinders',
 'enginesize',
 'horsepower',
 'citympg',
 'highwaympg',
 'price')

#### **Extracting the shape of the array**

In [4]:
print("The shape of the array is: ", data.shape)

The shape of the array is:  (205,)
('audi', 'gas', 'four', 'sedan', 'fwd', 105.8, 192.7, 71.4, 55.7, 'five', 136, 110, 19, 25, 17710)


#### **T2) What do you see?**
* How many entries does the array have?
* What does each entry include? 
* Hint: Print the elements of an entry


The array has 205 entries
Each entry includes a make, fueltype, numofdoors, boydstyle, drivewheels, wheelbase, length, width, height, numofcylinders,enginesize, horsepower, citympg, highwaympg and price.

#### **Extracting the entries of a column given its name**

* By specifying the name of a column, you can get all the entries within the array for this column (reminder: you are using Structured Arrays)


In [5]:
print(data['make'])

['alfa-romero' 'alfa-romero' 'alfa-romero' 'audi' 'audi' 'audi' 'audi'
 'audi' 'audi' 'audi' 'bmw' 'bmw' 'bmw' 'bmw' 'bmw' 'bmw' 'bmw' 'bmw'
 'chevrolet' 'chevrolet' 'chevrolet' 'dodge' 'dodge' 'dodge' 'dodge'
 'dodge' 'dodge' 'dodge' 'dodge' 'dodge' 'honda' 'honda' 'honda' 'honda'
 'honda' 'honda' 'honda' 'honda' 'honda' 'honda' 'honda' 'honda' 'honda'
 'isuzu' 'isuzu' 'isuzu' 'isuzu' 'jaguar' 'jaguar' 'jaguar' 'mazda'
 'mazda' 'mazda' 'mazda' 'mazda' 'mazda' 'mazda' 'mazda' 'mazda' 'mazda'
 'mazda' 'mazda' 'mazda' 'mazda' 'mazda' 'mazda' 'mazda' 'mercedes-benz'
 'mercedes-benz' 'mercedes-benz' 'mercedes-benz' 'mercedes-benz'
 'mercedes-benz' 'mercedes-benz' 'mercedes-benz' 'mercury' 'mitsubishi'
 'mitsubishi' 'mitsubishi' 'mitsubishi' 'mitsubishi' 'mitsubishi'
 'mitsubishi' 'mitsubishi' 'mitsubishi' 'mitsubishi' 'mitsubishi'
 'mitsubishi' 'mitsubishi' 'nissan' 'nissan' 'nissan' 'nissan' 'nissan'
 'nissan' 'nissan' 'nissan' 'nissan' 'nissan' 'nissan' 'nissan' 'nissan'
 'nissan' 'nissa

#### **T3) Extract the bodystyles within the dataset**


In [6]:
print(data['bodystyle'])

['convertible' 'convertible' 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan'
 'wagon' 'sedan' 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan'
 'sedan' 'sedan' 'sedan' 'hatchback' 'hatchback' 'sedan' 'hatchback'
 'hatchback' 'hatchback' 'hatchback' 'sedan' 'sedan' 'sedan' 'wagon'
 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback'
 'sedan' 'wagon' 'hatchback' 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan'
 'sedan' 'sedan' 'sedan' 'hatchback' 'sedan' 'sedan' 'sedan' 'hatchback'
 'hatchback' 'hatchback' 'sedan' 'sedan' 'hatchback' 'hatchback'
 'hatchback' 'hatchback' 'hatchback' 'sedan' 'hatchback' 'sedan' 'sedan'
 'hatchback' 'sedan' 'sedan' 'sedan' 'wagon' 'hardtop' 'sedan' 'sedan'
 'convertible' 'sedan' 'hardtop' 'hatchback' 'hatchback' 'hatchback'
 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback' 'hatchback'
 'hatchback' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan' 'sedan'
 'sedan' 'wagon' 'sedan' 'hatchback' 'sedan' 'wagon' 'hardtop' 'hatchback'
 'seda

### How do the car prices look like?


#### **T4) Calculate the range of car prices for the entire dataset**


In [17]:
print(np.ptp(data['price']))

40282


#### **T5) Calculate the min, max, mean and median prices of the cars**


In [19]:
print(np.min(data['price']))
print(np.max(data['price']))
print(np.mean(data['price']))
print(np.median(data['price']))

5118
45400
13300.239024390245
10345.0


#### **T6) Considering the values calculated above, what insights can you extract? Where do you think the majority of car prices will be clustered?**


The min is 5118 and range is 40282. The median, 10345 is roughly 5000 away from the min compared to 25000 away from the max so most prices are probably close to 5118.

#### **T7) Write code to calculate the standard deviation for the car prices. Then use the corresponding NumPy function to confirm the correctness of your calculation**



In [45]:
var = 0
for i in range(len(data)):
    var += (data[i]['price'] - np.mean(data['price']))**2
var = var/len(data)
print(np.sqrt(var))
print(np.std(data['price']))

7969.541401038539
7969.54140103854


#### **T8) Find the details of cars with the smallest and largest car volumes**
* Hint: see how to calculate the volume of a car https://info.japanesecartrade.com/content-item/297-what-is-m3-cubic-meter-size-of-a-vehicle


In [52]:
minM3 = 0
minIndex = 0
maxM3 = 0
maxIndex = 0
for i in range(len(data)):
    M3 = data[i]['width'] * data[i]['length'] * data[i]['height']
    if minM3 == 0 or M3 < minM3:
        minM3 = M3
        minIndex = i
    if M3 > maxM3:
        maxM3 = M3
        maxIndex = i
print(data[minIndex])
print('has the minimum M3 of', minM3)
print(data[maxIndex])
print('has the maximum M3 of', maxM3)

('chevrolet', 'gas', 'two', 'hatchback', 'fwd', 88.4, 141.1, 60.3, 53.2, 'three', 61, 48, 47, 53, 5151)
has the minimum M3 of 452643.16
('mercedes-benz', 'gas', 'four', 'sedan', 'rwd', 120.9, 208.1, 71.7, 56.7, 'eight', 308, 184, 14, 16, 40960)
has the maximum M3 of 846007.6


#### **T9) Find the different types of bodystyles for the cars in the dataset**

* Hint: You may want to check: https://numpy.org/doc/stable/reference/generated/numpy.unique.html

In [53]:
print(np.unique(data['bodystyle']))

['convertible' 'hardtop' 'hatchback' 'sedan' 'wagon']


#### **T10) Find the number of different car *brands* (makes)**


In [55]:
print(len(np.unique(data['make'])))

22


#### **T11) Find the engine size and the horsepower for the most and least efficient cars when driven in the city and the highway (i.e., the cars with the smallest and largest difference in fuel consumption when driven in the city and the highway)**

In [58]:
mostEff = 0
mostIndex = 0
leastEff = 0
leastIndex = 0
for i in range(len(data)):
    Eff = data[i]['citympg'] - data[i]['highwaympg']
    if leastEff == 0 or Eff < leastEff:
        leastEff = Eff
        leastIndex = i
    if Eff > mostEff:
        mostEff = Eff
        mostEff = i
print('The least efficient car has and engine size of', data[leastIndex]['enginesize'], 'and a horsepower of', data[leastIndex]['horsepower'])
print('The most efficient car has and engine size of', data[mostIndex]['enginesize'], 'and a horsepower of', data[mostIndex]['horsepower'])

The most efficient car has and engine size of 203 and a horsepower of 288
The most efficient car has and engine size of 130 and a horsepower of 111


#### **T12) Find the make with the largest number of cars and how many they are**

In [62]:
mostCars = 0
mostIndex = 0
cars = 0
for i in range(len(np.unique(data['make']))):
    for j in range(len(data)):
        if data[j]['make'] == data[i]['make'] :
               cars += 1
    if cars > mostCars:
        mostCars = cars
        mostIndex = i
print('The make with the largest number of cars is', data[mostIndex]['make'], 'with', mostCars, 'cars')

The make with the largest number of cars is dodge with 140 cars


#### **T13) Find how many cars have a wheel base greater than 100**

* Hint: See https://jakevdp.github.io/PythonDataScienceHandbook/02.06-boolean-arrays-and-masks.html

In [64]:
cars = 0
for i in range(len(data)):
    if data[i]['wheelbase'] > 100:
        cars += 1
print(cars, 'cars have a wheel base greater than 100')

63 cars have a wheel base greater than 100


#### **T14) Find if there are any convertible cars that cost less than £15000**

* Hint: See https://jakevdp.github.io/PythonDataScienceHandbook/02.06-boolean-arrays-and-masks.html

In [71]:
boolean = False
for i in range(len(data)):
    if data[i]['bodystyle'] == 'converitble' and data[i][price] < 15000.00:
        boolean = True
        break
if boolean is True:
    print('There is convertible cars that are less than £15000')
else:
    print('There is no convertible cars that are less than £15000')

There is no convertible cars that are less than £15000


#### **T15) Calculate the interquartile range for the price of all cars**

In [79]:
Q1 = np.percentile(data['price'], 25)
Q3 = np.percentile(data['price'], 75)
IQ = Q3-Q1
print("Interquartile range of prices is", IQ)

Interquartile range of prices is 8715.0


#### **T16) Calculate the 50th percentile range for the horsepower of all cars. Which value is the 50th percentile equal to?**

In [82]:
print('The 50th percentile range is for horsepower is', np.percentile(data['horsepower'], 75) - np.percentile(data['horsepower'], 25))
print('The 50th percentile of horsepower is', np.percentile(data['horsepower'], 50))

The 50th percentile range is for horsepower is 46.0
The 50th percentile of horsepower is 95.0


### Ideas for practicing further at home

* Find the engine and horsepower of 4wd cars
* Find whether diesel or gas cars are more efficient in the city/highway
* Any other analysis that you might could generate some useful insight.
